### PR TITLE
#9035 + #5936 Rebase

### PR DESCRIPTION
**Pull request by kylekatarnls**
Rebase the patch branch

### CODE REVIEW

File: app/views/projects/blob/_editor.html.haml [hunk 1]
```diff
@@ -17,7 +17,10 @@
 
   .file-content.code
     %pre.js-edit-mode-pane#editor
-      = params[:content] || local_assigns[:blob_data]
+      - content = params[:content] || local_assigns[:blob_data]
+      = content
+      - if content.length && 0 === content[content.length-1].index(/[\r\n]/)
```
Comment (stanhu) [inline app/views/projects/blob/_editor.html.haml]: I'm a bit confused how this works. There's an `if` clause here with a non-printing line.

Comment (kylekatarnls) [inline app/views/projects/blob/_editor.html.haml]: It do what this issues are about: preserve the trailing end of line (avoid the auto-trim in previous version and now avoid a line to be added automatically). In short word, you will not have the end of the file modified if you haven't added manually: if the last char is a new line, re-add it: https://github.com/gitlabhq/gitlabhq/pull/9422

Comment (stanhu) [inline app/views/projects/blob/_editor.html.haml]: I did some cursory debugging and found the right number of newlines get sent from the UI editor to the controller. Is the issue that the view is truncating the last newline if it's there?

Comment (kylekatarnls) [inline app/views/projects/blob/_editor.html.haml]: No, the other way, the last new line from the controller to the editor was lost.

**Comment by dosire** [comment]
This problem is solved in master.

**Comment by dosire** [comment]
See https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/2099

**Pull request closed without merging**